src/util/serializer.h

```diff
@@ -0,0 +1,818 @@
+      {
+        typename sett::value_type elt = create_elt();
+        serializer.serialize("value", elt);
+        field.insert(elt);
+      });
+    }
+    else
+    {
+      serializable_collection_writert<sett> writer(field);
+      write_array(name, writer);
+    }
+  }
+
+private:
+  /*******************************************************************\
+
+  Function: serializert::serialize_set
+
+  Template parameters:
+    sett:
+      The type of set to be serialized.
+      Must have a member type value_type that is serializable, has a default
+        constructor and can be inserted into the set.
+
+  Inputs:
+    name: The name of the field to serialize.
+    field: The field to serialize.
+
+  Outputs:
+
+  Purpose:
+    Helper function to serialize a set field.
+
+  \*******************************************************************/
+  template<typename sett>
+  void serialize_set(const char *name, sett &field)
+  {
+    serialize_set(name, field, [] () { return typename sett::value_type(); });
+  }
+
+public:
+  /*******************************************************************\
+
+  Function: serializert::serialize
+
+  Template parameters:
+    eltt:
+      The type of elements of the set.
+      Must be serializable and have a default constructor.
+
+  Inputs:
+    name: The name of the field to serialize.
+    field: The field to serialize.
+
+  Outputs:
+
+  Purpose:
+    Serializes a field of type std::set<eltt>.
+
+  \*******************************************************************/
+  template<typename eltt>
+  void serialize(const char *name, std::set<eltt> &field)
+  {
+    serialize_set(name, field);
+  }
+
+  // Serializing maps of serializable values
+private:
+  // This helper function can serialize any type of map (mapt) that has member
+  //  types key_type, mapped_type, value_type and reference, an insert method
+  //  that takes std::pair<mapt::key_type, mapt::mapped_type> and can be
+  //  enumerated to get elements of type mapt::reference from which a
+  //  serializable_pairt<mapt::value_type> can be created.
+  template<typename mapt>
+  void serialize_map(const char *name, mapt &field)
+  {
+    if(is_read)
+    {
+      read_array(
+        name,
+        [&field](serializert &serializer)
+      {
+        typedef std::pair<typename mapt::key_type, typename mapt::mapped_type>
+          pairt;
+        pairt key_value_pair;
+        serializable_pairt<pairt> wrapper(key_value_pair);
+        wrapper.serialize(serializer);
+        field.insert(key_value_pair);
+      });
+    }
+    else
+    {
+      typedef serializable_pairt<typename mapt::value_type> wrappert;
+      std::vector<wrappert> wrapper_vector;
+      for(typename mapt::reference elt : field)
+        wrapper_vector.emplace_back(elt);
+      serializable_obj_collection_writert<std::vector<wrappert>>
+        writer(wrapper_vector);
+      write_array(name, writer);
+    }
+  }
+
+public:
```

Comment:
put public section first so that I don't have to scroll down to see the interface of the class